### ISSUE TITLE
Hint marker text color is overwritten by other styles

### ISSUE THREAD

**Issue by hyperfekt** [philc/vimium#2946]
When a website includes some style that declares things a color with the `!important` keyword, the letters of the link hints will be affected, potentially making them very hard to read against the background.
Therefore I suggest declaring the hint marker's text color as `!important` by default, which will overrride the style applied by the website.

**Comment by smblott-github** [comment]
Do you have an example site for this issue?

**Comment by dglava** [comment]
I don't know if this is related, but I'm experiencing similar issues. On some websites (can only reproduce on Github.com so far), my background color style gets reset to what appears to be the default style.
I am using Vimium on Firefox 58.0.1

Hint markers on Github.com:
https://s9.postimg.org/mqmyclhrj/2018-02-08-09_09_36.png

Hint markers on other sites (according to what I set in the settings):
https://s9.postimg.org/87ftb6wcf/2018-02-08-09_13_23.png

My CSS for Vimium UI:
```
div > .vimiumHintMarker {
/* linkhint boxes */
background: #FFC542;
border: 1px solid #E3BE23;
}

div > .vimiumHintMarker span {
/* linkhint text */
color: black;
font-weight: bold;
font-size: 12px;
}

div > .vimiumHintMarker > .matchingCharacter {
}
```

**Comment by hyperfekt** [comment]
It might very well be that the issue only appears if the stylesheet is loaded by another extension.
I haven't managed to create a simple example, but using Stylish on Firefox with the ['My Twitter Dark' style](https://userstyles.org/styles/147945/my-twitter-dark) will have the effect I'm describing on twitter.com.

**Comment by ght** [comment]
TL;DR; I suggest adding `!important` to all properties in the default style to fix this and possible related issues for all users.

It makes Vimium more robust and compatible with other extensions which might insert styles into the page in yet unforseen ways, instead of requiring every other extension to make sure it is compatible (which won't happen, and it could be lots of wasted time debugging).

For example this issue also occurs when using user style sheets with [Stylus](https://github.com/openstyles/stylus) (for example [Wikipedia-Dark](https://github.com/StylishThemes/Wikipedia-Dark), where an issue about this was also reported StylishThemes/Wikipedia-Dark#52).

If there are any technical reasons against this, please share them. There is already a PR trying to fix this (#2533, opened June 2017), which adds `!important` to the color in `content_scripts/vimium.css`. I think the better approach would be to change the default style in [`lib/settings.coffee`](vimium/blob/1aeaa8d/lib/settings.coffee#L137-L152); I will open a PR for that in the next few days if there are no objections.

**Comment by gdh1995** [comment]
In face, the simplest code to fix this is to move LinkHints' `vimiumHintMarkerContainer` inside a shadow root.

I also suggest that we increases Vimium-supported minimum Chrome version to 53, and minimum Firefox version to 63, which means they ensure supports for Shadow DOM v1, and then Vimium can use `attachShadow(mode: "closed")` to create a "closed" ShadowRoot, so that no outside (host-page) CSS selectors can match its inner nodes.

**Comment by ght** [comment]
That would of course be even better, I haven't worked with Shadow DOM yet though, so I don't think I will be able to create a PR for that in the near future.

**Comment by philc** [comment]
@gdh1995 cool idea, much more robust than what we've been doing, and thus ultimately simpler. Would you like to give it a try?

**Comment by gdh1995** [comment]
We may add a class name to `<span>`s of `div.vimiumHintMarker`, and then a special `!important` for them is much direct.

I'm in work these day and can not afford enough time to finish such a big reworking.
